### PR TITLE
Fix jupyter io

### PR DESCRIPTION
**Pull request by stevenrbrandt**
Using Python's print routine for I/O redirection is problematic. In principle, one probably needs the GIL to do it. This branch uses linux pipes to redirect output (but only if we're running inside a notebook).

### CODE REVIEW

File: python/src/phylanx.cpp [hunk 1]
```diff
@@ -11,6 +11,9 @@
 #include <bindings/binding_helpers.hpp>
 
 #include <pybind11/pybind11.h>
+#include <pybind11/iostream.h>
+#include <pybind11/operators.h>
+#include <pybind11/stl.h>
```
Comment (hkaiser) [inline python/src/phylanx.cpp]: Why are these #include's necessary for this file?

**Comment by hkaiser** [comment]
Could you please explain why using the GIL would be problematic? While your solution might work it looks to be too complicated for a simple thing as getting output redirected to stdout.

Please provide some rationale for your design decision.

File: python/phylanx/ast/transformation.py [hunk 2]
```diff
@@ -320,10 +345,23 @@ def _UnaryOp(self, a):
...
+        sep = ""
+        for arg in ast.iter_child_nodes(a):
+            if arg.__class__.__name__ == "Load":
+                break
+            s += sep
+            s += self.recompile(arg)
+            sep = ","
+        s += ")"
+        return s
+
```
Comment (hkaiser) [inline python/phylanx/ast/transformation.py]: Is this change unrelated?

File: python/phylanx/ast/transformation.py [hunk 3]
```diff
@@ -451,6 +500,7 @@ def __init__(self, f):
 
         def __call__(self, *args):
             nargs = tuple(convert_to_phylanx_type(a) for a in args)
-            return et.eval(self.__physl_src__, *nargs)
+            with iod:
+                return et.eval(self.__physl_src__, *nargs)
```
Comment (hkaiser) [inline python/phylanx/ast/transformation.py]: Wouldn't this disable evaluating things anywhere except in Jupyter notebooks?

File: python/phylanx/ast/transformation.py [hunk 4]
```diff
@@ -418,6 +458,15 @@ def convert_to_phylanx_type(v):
         pass
     return v
 
+# Determine whether we are
+# running in a notebook or
+# not. If we are, use the
+# IORedirector.
+try:
+    get_ipython()
```
Comment (hkaiser) [inline python/phylanx/ast/transformation.py]: Does `get_ipython` really check whether we're running in a notebook. Wouldn't this call succeed in any IPython environment?

File: python/src/phylanx.cpp (hunk 1 above)
Comment (stevenrbrandt) [reply python/src/phylanx.cpp]: Sorry, let me clean up the unrelated stuff.

File: python/phylanx/ast/transformation.py (hunk 3 above)
Comment (hkaiser) [reply python/phylanx/ast/transformation.py]: Ahh sorry, I misread the code. Please disregard my comment.

File: python/phylanx/ast/transformation.py (hunk 4 above)
Comment (stevenrbrandt) [reply python/phylanx/ast/transformation.py]: Yes, it will also succeed in an IPython environment, but not from the regular command line.

Comment (hkaiser) [reply python/phylanx/ast/transformation.py]: Would this be a problem?

Comment (stevenrbrandt) [reply python/phylanx/ast/transformation.py]: It means IPython will use the IO redirection, which it doesn't need to do. I suppose I could check os.environ[_] and look for the string "jupyter" and/or "ipython". The latter feels more hacky to me.

**Pull request merged**